### ISSUE TITLE
Redesign to address focus/blur listeners - Mass slider's (number picker) dynamic aria-valuetext

### ISSUE THREAD

**Issue by terracoda** [phetsims/gravity-force-lab-basics#125]
@zepumph, apologies, maybe more detail than you need...

Due to the lack of focus and blur events with VoiceOver on iOS, @zepumph and @terracoda discussed two possible design changes for providing the comparative size/density descriptions for the masses.

For reference, currently, the third bullet for both mass 1, Blue sphere and mass 2, Red Sphere do not change when Constant Size is checked as no comparison is provided. Examples:
- ‪mass 1‬ is 2 billion kilograms, and 4 kilometers from ‪mass 2‬.
- ‪mass 2‬ is 4 billion kilograms, and 4 kilometers from ‪mass 1‬.

**Option 1 - use less verbose `aria-valuetext` all the time, and move comparison to mass sphere dynamic descriptions, 3rd bullet**
Moving the comparative description to the sphere descriptions requires that we have to dynamically swap between size and density comparisons. The `aria-valuetext` then requires no special logic on a focus event, and always just provides the current mass. 
- Constant Size not checked:
  - ‪mass 1‬ is 2 billion kilograms, _half the size of mass 2,_ and 4 kilometers _away‬_ .
  - ‪mass 2‬ is 4 billion kilograms, _twice the size of mass 1,_ and 4 kilometers _away_ .
- Constant Size checked:
  - ‪mass 1‬ is 2 billion kilograms, _half as dense as mass 2,_ and 4 kilometers _away_‬ .
  - ‪mass 2‬ is 4 billion kilograms, _twice as dense as mass 1,_ and 4 kilometers _away_‬ .

@zepumph, note I change the end of the distance string to "away". We didn't discuss this part at the meeting. I think it sounds more natural and hopefully doesn't complicate things too much. I italicized the additions/changes.

**Option 2 - use the more verbose `aria-valuetext` containing the comparisons all the time, and leave mass sphere description as is**
In this case, the `aria-valuetext` will always be verbose (on focus and when changed).
With Constant Size not checked:
- Focus: 2 billion kilograms, half the size of ‪mass 2‬, Mass 1, number spinner
- Changed: 3 billion kilograms, smaller but comparable to ‪mass 2‬
With Constant Size checked

Option 2 only possible because the comparison description happens to come second in the `aria-valuetext`.
- Focus: 2 billion kilograms, half as dense as ‪mass 2‬, Mass 1, number spinner
- Changed: 3 billion kilograms, less dense but comparable to ‪mass 2‬

@zepumph, I think I would like 2 versions one for option 1 and one with option 2 to try with the consultants.

**Comment by terracoda** [comment]
@zepumph, actually you might only have time to do a single option, so just do option 2.

I need a version for May 7th in the morning, (see #122), I'll need the change ported to the sonification wrapper, too. So this might be tight.

**Comment by terracoda** [comment]
This issue is related to https://github.com/phetsims/a11y-research/issues/138

**Comment by terracoda** [comment]
This issue address requirements for a11y mobile issues #119.

**Comment by zepumph** [comment]
Option 2 has been implemented above, let me know what you think!

**Comment by zepumph** [comment]
I noticed that right now the aria-valuetext doesn't update when the constant-size checkbox is toggled. That can mean a buggy value read out on focus.

**Comment by terracoda** [comment]
@zepumph, I like hearing the comparisons for size or density after the value for mass. I think this will work.

Yes, I hear the stale `aria-valuetext` upon focus of a mass control just after the Constant Size checkbox has been checked or unchecked. I'll, mention this to the consultants of they notice this issue.

@zepumph, is it reasonable to assume, we could handle this bug by forcing a reset on both mass controls when the Constant Size checkbox is checked? Or is that a bad idea? It's likely very sim specific and not very generalisable.

**Comment by terracoda** [comment]
@zepumph, and just to follow-up on that last idea, we already have to swap the help text paragraph when the Constant Size checkbox is checked or unchecked, so maybe we refreshing the `aria-valuetext` on teh mass controls at the same time would be ok?

**Comment by terracoda** [comment]
@zepumph, any further changes to the mass controls can wait until after this weeks interviews, so changing the priority to medium.

**Comment by zepumph** [comment]
> It's likely very sim specific and not very generalisable.

Yes that is sim specific, though that isn't necessarily a deal breaker. I just wish there was a better way to handle edge cases like these. Let's discuss more at our meeting.

**Comment by zepumph** [comment]
@terracoda will take a look at this from a design perspective. From our meeting, we are leaning on removing the comparisons from the aria valuetext (since they can become out of sync when moving the other mass), and instead move that perhaps to the sphere descriptions. @terracoda also thought though that that may be too verbose, because that info is already in the screen summary. She will have a look and assign it back to me.

**Comment by zepumph** [comment]
The other option would be keep the "verbose" comparison valuetext in the mass controls, and to update the aria-valuetext each time (1) this mass control changes, (2) the other mass control changes, (3) the constant size checkbox changes. Thus treating this aria-valuetext more like a dynamic description than a value dependent on the aria-valuetext. We will only go down this road if we need to.

**Comment by terracoda** [comment]
@zepumph, due to issues with stale `aria-valuetext`, I think it is best to remove the comparative descriptions all together from the mass 1 and mass 2 number spinners.

To get this sim into dev testing as close to on time as possible, @emily-phet and I decided it is less problematic to implement option 1 as described below and in  in https://github.com/phetsims/gravity-force-lab-basics/issues/125#issue-439083669

We can consider a maintenance release of this sim once stale `aria-valuetext` is fully address in #128. We can also work on #128 while finishing off descriptions for GFL Regular.

**To Re-iterate Option 1:**
**Use less verbose `aria-valuetext` all the time, and move the comparative descriptions for mass and density to the mass sphere dynamic descriptions in the 3rd bullet**

Moving the comparative description to the sphere descriptions requires that we have to dynamically swap between size and density comparisons. The `aria-valuetext` then requires no special logic on a focus event, and always just provides the current mass, which is the same whether or not Constant Size is checked. 
- The third bullet in the mass sphere description with Constant Size not checked:
  - ‪mass 1‬ is 2 billion kilograms, _half the size of mass 2,_ and 4 kilometers _away‬_ .
  - ‪mass 2‬ is 4 billion kilograms, _twice the size of mass 1,_ and 4 kilometers _away_ .
- The third bullet with Constant Size checked:
  - ‪mass 1‬ is 2 billion kilograms, _half as dense as mass 2,_ and 4 kilometers _away_‬ .
  - ‪mass 2‬ is 4 billion kilograms, _twice as dense as mass 1,_ and 4 kilometers _away_‬ .

@zepumph, please not that the italics highlight the new inserted comparative phrase and a small wording change at the end of the entire bullet.

I'll assign to you as I think this is straight forward without examples in the design document, but that said i will update the design doc.

**Comment by terracoda** [comment]
@zepumph, my solution in the previous comment doesn't work if distance is not checked.
Assigning back to myself!!! Arrgh!

**Comment by terracoda** [comment]
@zepumph, ok after much toil and trouble I couldn't re-work the third-bullet phrase to not mention the second mass twice and also not create a need to change the parameter strings.

I feel that the simplest and best solution is to add the word “_**also**_” (emphasis mine) to the unchanging part of the second half of the phrase. When we do this, I think it sounds ok, and most importantly, there is no need for a new set of slightly changed parameters (YEAH!).

Note: I have only put the newly added parameter in curly braces. The rest of the parameters are not marked.

**When Distance checked**
- mass 1‬ is 2 billion kilograms, {{half the size of}} mass 2, and **_also_** 4 kilometer from ‪mass 2‬.

**When Distance not checked**
- mass 1‬ is 2 billion kilograms, {{half the size of}} mass 2, and **_also_** close to ‪mass 2‬.

**When Constant Size checked:**
- mass 1 is 2 billion kilograms, {{half as dense as}} mass 2, and **_also_** 4 kilometers from mass 2.

**When Constant Size checked & Distance not checked:**
- mass 1 is 2 billion kilograms, {{half as dense as}} mass 2, and **_also_** close to mass 2.

Assigning back to you @zepumph to remove comparative description in mass `aria-valuetex` and to rebuild the third bullet on each mass sphere dynamic state description.

**Comment by zepumph** [comment]
Implemented above. Now the mass control aria-valuetext is just the mass and unit (for both GFLB and REGULAR). Also in GFLB the PDOM Node's third bullet has been updated to your spec (not in REGULAR). 

Please review @terracoda.

**Comment by terracoda** [comment]
@zepumph, the mass control `aria-valuetext` and the content of the third bullet sound good.
While verifying this issue I noticed a bug in the `aria-valuetext`. I heard movement of masses described when a change was not cause movement. Opening a new issue for that.
